### ISSUE TITLE
Flutter driver --use-existing-app error

### ISSUE THREAD

**Issue by antoninobajeli** [flutter/flutter#22550]
I wrote an integration test that is currently working if launched in a standard way.

> **i.e. flutter drive --target=test_driver/user_list_scrolling.dart **

When I try to launch it specifying to use the installed App ( --use-existing-app param) I get the error.

This is the command:
> **flutter drive --target=test_driver/user_list_scrolling.dart --use-existing-app=http://127.0.0.1:56302**

This is the output :
```
Using device TA 1021.
Will connect to already running application instance.
00:00 +0: scrolling performance test (setUpAll)
[info ] FlutterDriver: Connecting to Flutter application at http://127.0.0.1:56302
[trace] FlutterDriver: Isolate found with number: 736100165
[trace] FlutterDriver: Isolate is not paused. Assuming application is ready.
00:00 +0 -1: scrolling performance test (setUpAll) [E]
  DriverError: Failed to fulfill GetHealth due to remote error
  Original error: JSON-RPC error -32601 (method not found): Method not found
  Original stack trace:
  package:json_rpc_2/src/client.dart 110:64        Client.sendRequest
  package:json_rpc_2/src/peer.dart 68:15           Peer.sendRequest
  package:vm_service_client/src/scope.dart 64:23   Scope.sendRequestRaw
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1053:19                     _CustomZone.registerUnaryCallback
  dart:async/runtime/libasync_patch.dart 77:23     _asyncThenWrapperHelper
  package:flutter_driver/src/driver/driver.dart    FlutterDriver.connect
  test_driver\user_list_scrolling_test.dart 16:36  main.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1053:19                     _CustomZone.registerUnaryCallback
  dart:async/zone.dart 967:22                      _CustomZone.bindUnaryCallbackGuarded
  dart:async/future.dart 524:34                    Future.doWhile
  dart:async/future.dart 484:12                    Future.forEach
  package:test/src/backend/declarer.dart 291:36    Declarer._setUpAll.<fn>.<fn>
  dart:async/zone.dart 1124:13                     _rootRun
  dart:async/zone.dart 1021:19                     _CustomZone.run
  dart:async/zone.dart 1516:10                     _runZoned
  dart:async/zone.dart 1463:12                     runZoned
  package:test/src/backend/declarer.dart 291:14    Declarer._setUpAll.<fn>
  package:test/src/backend/invoker.dart 403:25     Invoker._onRun.<fn>.<fn>.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1045:19                     _CustomZone.registerCallback
  dart:async/zone.dart 962:22                      _CustomZone.bindCallbackGuarded
  dart:async/timer.dart 52:45                      new Timer
  dart:async/timer.dart 87:9                       Timer.run
  dart:async/future.dart 174:11                    new Future
  package:test/src/backend/invoker.dart 402:15     Invoker._onRun.<fn>.<fn>.<fn>


  package:flutter_driver/src/driver/driver.dart 359:7   FlutterDriver._sendCommand
  ===== asynchronous gap ===========================
  dart:async/future_impl.dart 22:43                     _Completer.completeError
  dart:async/runtime/libasync_patch.dart 40:18          _AsyncAwaitCompleter.completeError
  package:flutter_driver/src/driver/driver.dart         FlutterDriver._sendCommand
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1062:19                          _CustomZone.registerBinaryCallback
  dart:async/runtime/libasync_patch.dart 86:23          _asyncErrorWrapperHelper
  package:flutter_driver/src/driver/driver.dart         FlutterDriver._sendCommand
  package:flutter_driver/src/driver/driver.dart 382:34  FlutterDriver.checkHealth
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1053:19                          _CustomZone.registerUnaryCallback
  dart:async/runtime/libasync_patch.dart 77:23          _asyncThenWrapperHelper
  package:flutter_driver/src/driver/driver.dart         FlutterDriver.connect
  test_driver\user_list_scrolling_test.dart 16:36       main.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1053:19                          _CustomZone.registerUnaryCallback
  dart:async/zone.dart 967:22                           _CustomZone.bindUnaryCallbackGuarded
  dart:async/future.dart 524:34                         Future.doWhile
  dart:async/future.dart 484:12                         Future.forEach
  package:test/src/backend/declarer.dart 291:36         Declarer._setUpAll.<fn>.<fn>
  dart:async/zone.dart 1124:13                          _rootRun
  dart:async/zone.dart 1021:19                          _CustomZone.run
  dart:async/zone.dart 1516:10                          _runZoned
  dart:async/zone.dart 1463:12                          runZoned
  package:test/src/backend/declarer.dart 291:14         Declarer._setUpAll.<fn>
  package:test/src/backend/invoker.dart 403:25          Invoker._onRun.<fn>.<fn>.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async/zone.dart 1045:19                          _CustomZone.registerCallback
  dart:async/zone.dart 962:22                           _CustomZone.bindCallbackGuarded
  dart:async/timer.dart 52:45                           new Timer
  dart:async/timer.dart 87:9                            Timer.run
  dart:async/future.dart 174:11                         new Future
  package:test/src/backend/invoker.dart 402:15          Invoker._onRun.<fn>.<fn>.<fn>

00:00 +0 -1: scrolling performance test (tearDownAll)
Unhandled exception:
Dummy exception to set exit code.
#0      _rootHandleUncaughtError.<anonymous closure> (dart:async/zone.dart:1112:29)
#1      _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#2      _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#3      _Timer._runTimers (dart:isolate/runtime/libtimer_impl.dart:391:30)
#4      _Timer._handleMessage (dart:isolate/runtime/libtimer_impl.dart:416:5)
#5      _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)
00:00 +0 -1: Some tests failed.

Leaving the application running.
Driver tests failed: 255
```

This is the Flutter Doctor -v :
```
[√] Flutter (Channel beta, v0.8.2, on Microsoft Windows [Versione 10.0.17134.285], locale it-IT)
    • Flutter version 0.8.2 at c:\src\flutter
    • Framework revision 5ab9e70727 (4 weeks ago), 2018-09-07 12:33:05 -0700
    • Engine revision 58a1894a1c
    • Dart version 2.1.0-dev.3.1.flutter-760a9690c2

[√] Android toolchain - develop for Android devices (Android SDK 28.0.2)
    • Android SDK at C:\Users\abajeli\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[√] Android Studio (version 3.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 29.0.2
    • Dart plugin version 181.5616
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[√] Connected devices (1 available)
    • TA 1021 • PLEGAR17A1802749 • android-arm64 • Android 8.1.0 (API 27)

• No issues found!
```

**Comment by zoechi** [comment]
Is the existing app running?
Is the existing app a debug build?

**Comment by AGoodnight** [comment]
I have the same issue, I run
```flutter run```
then try to connect to an already running instance with the plugin method 'connect()'
```flutter drive --target=test_driver/integration.dart --use-existing-app=http://127.0.0.1:63749/```

I get the URL for the '--use-existing-app' from the launched application from 'flutter run'
```An Observatory debugger and profiler on Android SDK built for x86 64 is available at: http://127.0.0.1:63749/```

if I run it without setting '--use-existing-app' it runs fine.

I am unsure if it is a debug build or a regular build.

**Comment by ykorzikowski** [comment]
Same error here. I followed this tutorial: https://medium.com/flutter-community/hot-reload-for-flutter-integration-tests-e0478b63bd54

```
Yannik@workaholic:tiny_release $ flutter doctor -v
[✓] Flutter (Channel stable, v1.2.1, on Mac OS X 10.14.3 18D109, locale de-DE)
    • Flutter version 1.2.1 at /usr/local/Cellar/flutter/1.0.0
    • Framework revision 8661d8aecd (3 weeks ago), 2019-02-14 19:19:53 -0800
    • Engine revision 3757390fa4
    • Dart version 2.1.2 (build 2.1.2-dev.0.0 0a7dcf17eb)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/Yannik/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Users/Yannik/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/182.5264788/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 1.9.4
    • CocoaPods version 1.5.3

[!] Android Studio (version 3.3)
    • Android Studio at /Users/Yannik/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/182.5264788/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1248-b01)

[✓] IntelliJ IDEA Ultimate Edition (version 2018.3.5)
    • IntelliJ at /Users/Yannik/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 33.3.2
    • Dart plugin version 183.5912.23

[✓] Connected device (2 available)
    • Android SDK built for x86 • emulator-5554                            • android-x86 • Android 8.1.0 (API 27) (emulator)
    • Yanniks iPad (4)          • abc123 • ios         • iOS 12.1.4

! Doctor found issues in 1 category.

```

**Comment by lukaszciastko** [comment]
You are getting this error, because the application that you run doesn't call `enableFlutterDriverExtension()` in its `main` function.

The `flutter run` command will always run the app in `main.dart`. It's likely that your test app is in another file (e.g. `test_driver/integration.dart`). In which case, if everything is setup correctly, you should run:

`flutter run -t test_driver/integration.dart`

And only then you can run your tests:

`flutter drive --target=test_driver/integration.dart --use-existing-app=http://127.0.0.1:63749/`

I assume your `integration.dart` file calls `main()` to start the app. If it doesn't - it should; along calling `enableFlutterDriverExtension()`.

**Comment by bitsydarel** [comment]
I have the same issue even without use an existing application

flutter drive --debug --target=test_driver/app.dart
Using device Android SDK built for x86 64.
Starting application: test_driver/app.dart
Initializing gradle...                                              9.0s
Resolving dependencies...                                          45.6s
"build/app/outputs/apk/app.apk" does not exist.
Running Gradle task 'assembleDebug'...                             45.4s
Built build/app/outputs/apk/debug/app-debug.apk.
Installing build/app/outputs/apk/app.apk...                       470.9s (!)
I/flutter ( 6258): Observatory listening on http://127.0.0.1:46482/DN8mzFHc1S4=/
00:00 [32m+0[0m: Counter App (setUpAll)[0m
[info ] FlutterDriver: Connecting to Flutter application at http://127.0.0.1:37995/DN8mzFHc1S4=/
[trace] FlutterDriver: Isolate found with number: 3031536254716739
[warning] FlutterDriver: Unknown pause event type VMNoneEvent. Assuming application is ready.
00:12 [32m+0[0m[31m -1[0m: Counter App (setUpAll) [1m[31m[E][0m[0m
  JSON-RPC error -32601 (method not found): Method not found
  package:json_rpc_2/src/client.dart 110:64          Client.sendRequest

**Comment by bitsydarel** [comment]
And this happen on the example available here
https://flutter.dev/docs/cookbook/testing/integration/introduction

**Comment by ericmartineau** [comment]
@lukaszciastko When debugging FlutterDriver, I noticed that it was choosing the wrong isolate to run against.  In `driver.dart:243`, the driver chooses `vm.isolates.first`, but (at least in my case), the isolates were returned in _reverse_ order, eg:

<img width="1642" alt="bug_drive" src="https://user-images.githubusercontent.com/244358/72449961-c6406900-3776-11ea-91ee-29d1e58baca3.png">

I temporarily fixed it by changing `driver.dart:243` to read `vm.isolates.firstWhere((isolate) => iosolate.name == "main")`

**Comment by ericmartineau** [comment]
@jmagman @iapicca - I figure this is an old ticket so it's not going to get a lot of attention - will you take a look at what I posted above and let me know what you think?

**Comment by jmagman** [comment]
@ericmartineau  Not knowing much about this at all, that seems like a reasonable enough change, but heads up that code was moved to `vmservice_driver.dart` with https://github.com/flutter/flutter/pull/47890.

Would you like to open a PR?  You're going to want to add a test for this behavior, too, or your PR probably won't be approved.  [Check out how to mock the VM isolates.](https://github.com/flutter/flutter/blob/master/packages/flutter_driver/test/flutter_driver_test.dart#L50).  See also https://github.com/flutter/packages/blob/master/CONTRIBUTING.md#contributing-code.

**Comment by markusaksli-nc** [comment]
This seems fixed when testing on the latest master `1.21.0-6.0.pre.41`. Attached a driver test to running app with no errors.

`flutter drive --target=test_driver/app.dart --use-existing-app=http://127.0.0.1:64118/um_YkeUFs2w=/`
<details>
 <summary>Used test</summary>

 ```
import 'package:flutter_driver/flutter_driver.dart';
import 'package:test/test.dart';

void main() {
  group('Driver Test', () {
    FlutterDriver driver;

    setUpAll(() async {
      driver = await FlutterDriver.connect();
    });

    tearDownAll(() async {
      if (driver != null) {
        driver.close();
      }
    });
    test('basic test', () async {
      await driver.tap(find.byValueKey('fab'));
      expect(await driver.getText(find.byValueKey('title')), "Sample");
    });
  });
}
```
</details>
<details>
 <summary>Logs</summary>

 ```
Will connect to already running application instance.
00:00 +0: Driver Test (setUpAll)

VMServiceFlutterDriver: Connecting to Flutter application at http://127.0.0.1:64118/um_YkeUFs2w=/
VMServiceFlutterDriver: Isolate found with number: 2227744505971291
VMServiceFlutterDriver: Isolate is not paused. Assuming application is ready.
VMServiceFlutterDriver: Connected to Flutter application.
00:00 +0: Driver Test basic test

00:00 +1: Driver Test (tearDownAll)

00:00 +1: All tests passed!

Leaving the application running.
```
</details>
<details>
 <summary>flutter doctor -v</summary>

 ```
[√] Flutter (Channel master, 1.21.0-6.0.pre.41, on Microsoft Windows [Version 10.0.18362.959], locale et-EE)
    • Flutter version 1.21.0-6.0.pre.41 at C:\Development\flutter_master
    • Framework revision 39fa00201d (5 hours ago), 2020-07-27 21:11:43 -0700
    • Engine revision f27729e97b
    • Dart version 2.10.0 (build 2.10.0-0.0.dev 24c7666def)


[√] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at C:\Users\Isa\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: C:\Users\Isa\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.6.4)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.6.30309.148
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.0)
    • Android Studio at C:\Users\Isa\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.47.2)
    • VS Code at C:\Users\Isa\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.12.2

[√] Connected device (5 available)
    • SM G950F (mobile)       • ce12171c51cc001c03 • android-arm64  • Android 9 (API 28)
    • sdk gphone x86 (mobile) • emulator-5554      • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)       • windows            • windows-x64    • Microsoft Windows [Version 10.0.18362.959]
    • Web Server (web)        • web-server         • web-javascript • Flutter Tools
    • Chrome (web)            • chrome             • web-javascript • Google Chrome 84.0.4147.89

• No issues found!
```
</details>

Closing as probably fixed. If you still encounter this issue, please comment with your `flutter doctor -v` and I'll reopen.
